qubes/storage/callback.py

```diff
@@ -0,0 +1,442 @@
+            cmd = self._cb_conf.get(cb)
+            args = [] #on_xyz callbacks should never receive arguments
+            if not cmd:
+                cmd = self._cb_conf.get('cmd')
+                args = [ self.name, self._cb_conf['bdriver'], cb, self._cb_cmd_arg, *cb_args ]
+            if cmd and cmd != '-':
+                args = filter(None, args)
+                args = ' '.join(quote(str(a)) for a in args)
+                cmd = ' '.join(filter(None, [cmd, args]))
+                log.info('callback driver executing (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, cmd))
+                res = subprocess.run(cmd, shell=True, check=True, executable='/bin/bash', stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+                #stdout & stderr are reported if the exit code check fails
+                log.debug('callback driver stdout (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, res.stdout))
+                log.debug('callback driver stderr (%s, %s %s): %s' % (self._cb_conf_id, cb, cb_args, res.stderr))
+                if self._cb_conf.get('signal_back', False) is True:
+                    self._process_signals(res.stdout, log)
+
+    def _process_signals(self, out, log=logging.getLogger('qubes.storage.callback')):
+        '''Process any signals found inside a string.
+        :param out: String to check for signals. Each signal must be on a dedicated line.
+                    They are executed in the order they are found. Callbacks are not triggered.
+        '''
+        for line in out.splitlines():
+            if line == 'SIGNAL_setup':
+                log.info('callback driver processing SIGNAL_setup for %s' % self._cb_conf_id)
+                self.setup(callback=False)
+
+    def __del__(self):
+        s = super()
+        if hasattr(s, '__del__'):
+            s.__del__()
+
+    @property
+    def config(self):
+        return {
+            'name': self.name,
+            'driver': CallbackPool.driver,
+            'conf_id': self._cb_conf_id,
+        }
+
+    def destroy(self):
+        self._assertInitialized()
+        ret = self._cb_impl.destroy()
+        self._callback('on_destroy')
+        return ret
+
+    def init_volume(self, vm, volume_config):
+        return CallbackVolume(self, self._cb_impl.init_volume(vm, volume_config))
+
+    def setup(self, callback=True):
+        if callback:
+            self._callback('on_setup')
+        self._assertInitialized(callback=False) #setup is assumed to include initialization
+        return self._cb_impl.setup()
+
+    @property
+    def volumes(self):
+        for vol in self._cb_impl.volumes:
+            yield CallbackVolume(self, vol)
+
+    def list_volumes(self):
+        for vol in self._cb_impl.list_volumes():
+            yield CallbackVolume(self, vol)
+
+    def get_volume(self, vid):
+        return CallbackVolume(self, self._cb_impl.get_volume(vid))
+
+    def included_in(self, app):
+        if self._cb_requires_init:
```

Comment:
If it requires late init, you can also check if the init is already done - in which case call the actual implementation. This way, at least sometimes you'll get accurate data.

Reply:
`self._cb_requires_init` is set to `False` after storage initialization, so the `else`-path should handle that.